### PR TITLE
Allow submissions to use a different subset of testcases

### PR DESCRIPTION
**Pull request by kiritofeng**

### CODE REVIEW

File: dmoj/judge.py [hunk 1]
```diff
@@ -151,13 +151,13 @@ class InvocationCase(object):
...
     def _begin_grading(self, problem_id, language, source, time_limit, memory_limit, short_circuit, pretests_only,
-                       report=print):
+                       submission_number=0, report=print):
         submission_id = self.current_submission
         report(ansi_style('Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...'
                           % (problem_id, submission_id, language)))
 
         try:
-            problem = Problem(problem_id, time_limit, memory_limit, load_pretests_only=pretests_only)
+            problem = Problem(problem_id, time_limit, memory_limit, load_pretests_only=pretests_only, submission_number=submission_number)
```
Comment (quantum5) [inline dmoj/judge.py]: Line too long...

File: dmoj/packet.py [hunk 2]
```diff
@@ -205,7 +205,8 @@ def _receive_packet(self, packet):
                 float(packet['time-limit']),
                 int(packet['memory-limit']),
                 packet['short-circuit'],
-                packet['pretests-only']
+                packet['pretests-only'],
+                int(packet['submssion-number'] or '0')
```
Comment (quantum5) [inline dmoj/packet.py]: Huh? You mean `int(packet.get('submission-number') or 0)`?

File: dmoj/problem.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def __init__(self, problem_id, time_limit, memory_limit, load_pretests_only=Fals
         self.problem_data.archive = self._resolve_archive_files()
 
         self.is_pretested = load_pretests_only and 'pretest_test_cases' in self.config
-        self.cases = self._resolve_testcases(self.config['pretest_test_cases' if self.is_pretested else 'test_cases'])
+        self.cases = self._resolve_testcases(self.config['pretest_test_cases' if self.is_pretested else 'test_cases'], submission_number)
```
Comment (quantum5) [inline dmoj/problem.py]: Line too long.

File: dmoj/problem.py [hunk 4]
```diff
@@ -69,8 +69,11 @@ def _resolve_archive_files(self):
             return archive
         return None
 
-    def _resolve_testcases(self, cfg, batch_no=0):
+    def _resolve_testcases(self, cfg, submission_number=0, batch_no=0):
         cases = []
+        # do we use the same cases for all runs?
+        if 'default' in cfg:
+            cfg = cfg['run'+str(submission_number)] or cfg['default']
```
Comment (quantum5) [inline dmoj/problem.py]: Use `%d` and `%` formatting.

File: dmoj/graders/base.py [hunk 5]
```diff
@@ -1,15 +1,21 @@
...
         self.language = language
         self.problem = problem
         self.judge = judge
+        self.meta = ConfigNode(meta or {})
         self.binary = self._generate_binary()
         self._terminate_grading = False
         self._current_proc = None
+        self._batch_counter = 0
+        self._testcase_counter = 0
+        self.is_pretested = self.meta.get('pretests-only', False) and 'pretest_test_cases' in self.problem.config
```
Comment (quantum5) [inline dmoj/graders/base.py]: Huh? We control all the users of this...

Comment (kiritofeng) [reply dmoj/graders/base.py]: `is_pretested` is also called from `dmoj/judge.py`

**Pull request merged**